project_model_to_task/models/project.py

```diff
@@ -0,0 +1,105 @@
+            modules dependencies
+        """
+        models = self.env['ir.model'].search([])
+        return [(x.model, x.name) for x in models]
+
+    action_id = fields.Many2one(
+        'ir.actions.act_window', string="Action",
+        help="Action called to go to the original window.")
+    model_reference = fields.Reference(
+        selection='_authorised_models')
+    task_origin = fields.Char(compute='_get_origin')
+
+    @api.model
+    def default_get(self, fields):
+        vals = super(ProjectTask, self).default_get(fields)
+        if 'from_model' in self.env.context and 'from_id' in self.env.context:
+            vals['model_reference'] = '%s,%s' % (
+                self.env.context['from_model'],
+                self.env.context['from_id'])
+        if 'from_action' in self.env.context:
+            vals['action_id'] = self.env.context['from_action']
+        return vals
+
+    def goto_document(self):
+        self.ensure_one()
+        if self.model_reference:
+            action = {
+                'name': 'Task to original document',
+                'res_model': self.model_reference._name,
+                'res_id': self.model_reference.id,
+                'type': 'ir.actions.act_window',
+                'target': 'current',
+                'view_mode': 'form',
+            }
+            if self.action_id:
+                action['id'] = self.action_id.id
+                action['action_id'] = self.action_id.id
+                view = [x.view_id for x in self.action_id.view_ids
+                        if x.view_mode == 'form']
+                if view:
+                    view_ref = self.env['ir.model.data'].search(
+                        [('res_id', '=', view[0].id),
+                         ('model', '=', 'ir.ui.view')])
+                    if view_ref:
+                        action['context'] = {'form_view_ref': '%s.%s' % (
+                            view_ref.module, view_ref.name)}
+            return action
+        raise UserError(_(
+            "Field 'Task Origin' is not set.\n"
+            "Impossible to go to the original document."))
+
+
+class IrActionActWindows(models.Model):
+    _inherit = 'ir.actions.act_window'
+
+    def read(self, fields=None, load='_classic_read'):
+
+        def update_context(action):
+            action['context'] = safe_eval(action.get('context', '{}'))
+            action['context'].update({
+                'from_model': self.env.context.get('active_model'),
+                'from_id': self.env.context.get('active_id'),
+            })
+            if 'params' in self.env.context and 'action':
+                action['context'].update({
+                    'from_action': self.env.context['params'].get('action')})
+            if 'params' in self.env.context and 'action':
+                action['context'].update({
+                    'from_action': self.env.context['params'].get('action')})
+        res = super(IrActionActWindows, self).read(
```

Comment:
It would be nice to add an empty line here, to make it clear that the function declaration has ended